### PR TITLE
Add support for symlink and pnpm.

### PR DESCRIPTION
**Pull request by erossignon**
Add a mechanism to follow and track symlinks, making `pkg` work with PNPM and also package installed with `npm link` inside your `node_modules` folder.

Closes https://github.com/vercel/pkg/issues/941.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+image: node:14
```
Comment (leerob) [inline .gitlab-ci.yml:1]: Any reason why GitLab CI was added here?

Comment (erossignon) [reply .gitlab-ci.yml:1]: Yes, I designed/checked this PR in a private fork running on Gitlab prior to submitting it on GitHub... We can remove it now if you want. It is not harmful though.

Comment (erossignon) [reply .gitlab-ci.yml:1]: I have squashed the .gitlab-ci.yml commits into one

**Comment by PeledYuval** [comment]
We've *just* encountered this issue.

Every major build framework supports pnpm nowadays (webpack, gulp, create-react-app etc.) and we would love to have 'pkg' support it to - pnpm is a fantastic boon for us and we wouldn't want to choose one over the over.

I would really appreciate prioritizing this issue. Thanks to everybody for creating/approving this PR!

File: .gitignore [hunk 2]
```diff
@@ -1,8 +1,8 @@
 # build output
-/lib-es5
+lib-es5
 
 # dependencies
-/node_modules
+node_modules
```
Comment (leerob) [inline .gitignore:5]: Where these intentionally changed?

File: .gitlab-ci.yml (hunk 1 above)
Comment (leerob) [reply .gitlab-ci.yml:1]: Let's remove it 👍

File: lib/follow.js [hunk 3]
```diff
@@ -1,6 +1,8 @@
+
```
Comment (leerob) [inline lib/follow.js:1]:
```suggestion
```

File: lib/follow.js [hunk 4]
```diff
@@ -1,6 +1,8 @@
...
 import { core, sync } from 'resolve';
 import assert from 'assert';
 import fs from 'fs';
+import { normalizePath2 } from '../prelude/common';
```
Comment (leerob) [inline lib/follow.js:5]: Could we improve the name here?

File: lib/packer.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/* eslint-disable complexity */
+  /* eslint-disable complexity */
```
Comment (leerob) [inline lib/packer.js:1]:
```suggestion
/* eslint-disable complexity */
```

File: lib/walker.js [hunk 6]
```diff
@@ -430,9 +448,15 @@ class Walker {
 
     let newFile, failure;
 
+    const basedir = path.dirname(record.file);
+    // xx console.log(chalk.red("basedir "),basedir, derivative.alias);
```
Comment (leerob) [inline lib/walker.js:452]: Oops

File: lib/walker.js [hunk 7]
```diff
@@ -622,7 +658,10 @@ class Walker {
     for (const file of files) {
       if (/\.js$/.test(file)) {
         const name = file.slice(0, -3);
-        const config = require(path.join(dd, file));
+
+        const fff = path.join(dd, file);
+        //   fff = normalizePath(fff);
```
Comment (leerob) [inline lib/walker.js:663]: Oops

File: package-lock.json [hunk 8]
```diff
@@ -0,0 +1,4801 @@
+{
```
Comment (leerob) [inline package-lock.json:1]: This repo already has a yarn.lock - let's remove this?

File: package.json [hunk 9]
```diff
@@ -30,44 +30,44 @@
...
     "stream-meter": "^1.0.4"
   },
   "devDependencies": {
-    "@babel/cli": "^7.8.4",
-    "@babel/core": "^7.9.0",
-    "@babel/plugin-transform-async-to-generator": "^7.8.3",
-    "@babel/plugin-transform-runtime": "^7.9.0",
-    "@babel/preset-env": "^7.9.0",
-    "@babel/register": "^7.9.0",
+    "@babel/cli": "^7.13.10",
```
Comment (leerob) [inline package.json:48]: I'm not sure it's worth the risk to update all dependencies at the same time as adding a new feature. Thoughts?

File: prelude/bootstrap.js [hunk 10]
```diff
@@ -812,8 +812,10 @@ function payloadFileSync (pointer) {
   function readdirFromSnapshot (path_, isRoot, cb) {
     var cb2 = cb || rethrow;
     if (isRoot) return readdirRoot(path_, cb);
+
+
     var path = normalizePath(path_);
-    // console.log("readdirFromSnapshot", path);
+
```
Comment (leerob) [inline prelude/bootstrap.js:818]: Thanks!

File: lib/follow.js (hunk 4 above)
Comment (erossignon) [reply lib/follow.js:5]: yes, I propose : toNormalizedRealPath

File: package.json (hunk 9 above)
Comment (erossignon) [reply package.json:48]: All tests passes .... do we have enough coverage to gain confidence ?

File: prelude/bootstrap.js (hunk 10 above)
Comment (erossignon) [reply prelude/bootstrap.js:818]: There are some more ... in bootstrap.js

File: .gitlab-ci.yml (hunk 1 above)
Comment (erossignon) [reply .gitlab-ci.yml]: OK ,

File: package.json (hunk 9 above)
Comment (erossignon) [reply package.json]: I have move this chagne  to a dedicated pull request

**Comment by erossignon** [comment]
I could help as well.
I spent a significant amount of time getting into a deeper understanding of the virtual file system and script aggregation.
I am currently working on https://github.com/vercel/pkg/pull/1060 to support symlinks inside the virtual file system so that PKG can work when PNPM is used as a package manager.

**Comment by erossignon** [comment]
@leerob is there any else holding this pull request to be merged?

**Comment by leerob** [comment]
Would you mind resolving merge conflicts? 🙏

**Comment by erossignon** [comment]
For some reason, the 'eslint verification step" was removed in one recent commit.  
As a consequence, formatting has been mess up.
We might not like the eslint flavor, but it is important to have one, so that code is pushed consistently with the same formatting convention.

Anyway, I did fix the conflict issues, and I also restored the eslint step .. I also fix eslint error that was ignored in the test file (for some reason test files were removed from the lint phase as well)

**Comment by leerob** [comment]
Just merged ESLint - would you mind resolving conflicts again? Sorry about that! 🙏

**Comment by erossignon** [comment]
OK, remerged !

**Comment by robertsLando** [comment]
@leerob I think this could be merged

File: prelude/bootstrap.js [hunk 11]
```diff
@@ -837,9 +865,14 @@ function payloadFileSync(pointer) {
...
-  Dirent.prototype.isSocket = function isSocket() {
-    return false;
-  };
+  const noop = () => false;
+  Dirent.prototype.isBlockDevice = noop;
+  Dirent.prototype.isCharacterDevice = noop;
+  Dirent.prototype.isFIFO = noop;
+  Dirent.prototype.isSocket = noop;
+
+  Dirent.prototype.isSymbolicLink = (fileOrFolderName) =>
```
Comment (robertsLando) [inline prelude/bootstrap.js:874]: @erossignon Isn't this overrided by line 877?

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: prelude/bootstrap.js [hunk 12]
```diff
@@ -1291,10 +1319,6 @@ function payloadFileSync(pointer) {
     // For newer node versions (after https://github.com/nodejs/node/pull/33229 ):
     // Returns an array [string, boolean].
     //
-    var returnArray =
-      (NODE_VERSION_MAJOR === 12 && NODE_VERSION_MINOR >= 19) ||
-      (NODE_VERSION_MAJOR === 14 && NODE_VERSION_MINOR >= 5) ||
-      NODE_VERSION_MAJOR >= 15;
```
Comment (robertsLando) [inline prelude/bootstrap.js:1322]: @erossignon Why did you remove this?

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by leerob** [comment]
Lmk when ready to merge, I'll wait for your feedback above.

File: .eslintignore [hunk 13]
```diff
@@ -1,2 +1,2 @@
 /lib-es5
-/test
\ No newline at end of file
+node_modules
```
Comment (leerob) [inline .eslintignore:2]: Why did we ignore test to begin with?

**Comment by robertsLando** [comment]
This is huge and goes out of my understanding on some points, overall seems good except the points in the review. If tests are passing it means it isn't breaking anything. The feature it introduce is so important so we should take care of this

File: prelude/bootstrap.js [hunk 14]
```diff
@@ -1291,10 +1319,6 @@ function payloadFileSync(pointer) {
     // For newer node versions (after https://github.com/nodejs/node/pull/33229 ):
     // Returns an array [string, boolean].
     //
-    var returnArray =
```
Comment (jesec) [inline prelude/bootstrap.js:1294]: ?? Why this is removed?

File: prelude/bootstrap.js [hunk 15]
```diff
@@ -1308,15 +1332,17 @@ function payloadFileSync(pointer) {
...
+    if (!entity) {
+      return undefined;
+    }
+
     var entityContent = entity[STORE_CONTENT];
-    if (!entityContent) return returnArray ? [undefined, false] : undefined;
-    return returnArray
-      ? [payloadFileSync(entityContent).toString(), true]
-      : payloadFileSync(entityContent).toString();
+    if (!entityContent) {
```
Comment (jesec) [inline prelude/bootstrap.js:1342]: Is there a mismerge here?

File: prelude/bootstrap.js [hunk 16]
```diff
@@ -34,7 +38,6 @@ const {
 
 var FLAG_ENABLE_PROJECT = false;
 var NODE_VERSION_MAJOR = process.version.match(/^v(\d+)/)[1] | 0;
-var NODE_VERSION_MINOR = process.version.match(/^v\d+.(\d+)/)[1] | 0;
```
Comment (jesec) [inline prelude/bootstrap.js:37]: ??

**Review by jesec — DISMISSED** [review]

File: prelude/bootstrap.js (hunk 12 above)
Comment (erossignon) [reply prelude/bootstrap.js:1322]: Because it is not in used

File: .eslintignore (hunk 13 above)
Comment (erossignon) [reply .eslintignore:2]: Yes, make sense, test shall be also linted

File: prelude/bootstrap.js (hunk 14 above)
Comment (erossignon) [reply prelude/bootstrap.js:1294]: Because linter pointed out that the returnArray is not used anywhere later on

File: prelude/bootstrap.js (hunk 16 above)
Comment (erossignon) [reply prelude/bootstrap.js:37]: OK, merge issue.. let me fix it

File: prelude/bootstrap.js (hunk 15 above)
Comment (erossignon) [reply prelude/bootstrap.js:1342]: yep

File: prelude/bootstrap.js (hunk 14 above)
Comment (erossignon) [reply prelude/bootstrap.js:1294]: miss merge..

File: prelude/bootstrap.js (hunk 11 above)
Comment (erossignon) [reply prelude/bootstrap.js:875]: Right, well spotted.

Comment (erossignon) [reply prelude/bootstrap.js:877]: yes, some more refactoring was also possible

File: lib/packer.js [hunk 17]
```diff
@@ -88,68 +94,54 @@ export default function packer({ records, entrypoint, bytecode }) {
...
-            delete value.birthtimeMs;
-            // non-date
-            delete value.blksize;
-            delete value.blocks;
-            delete value.dev;
-            delete value.gid;
-            delete value.ino;
-            delete value.nlink;
-            delete value.rdev;
-            delete value.uid;
```
Comment (robertsLando) [inline lib/packer.js:115]: Why?

File: prelude/diagnostic.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+      );
+      return f.apply(this, args1);
+    };
+  }
+  if (process.env.DEBUG_PKG) {
+    console.log('------------------------------- virtual file system');
+    const startFolder = win32 ? 'C:\\snapshot' : '/snapshot';
+    const totalSize = dumpLevel(startFolder, 2);
+    console.log('Total size = ', totalSize);
+    wrap(fs, 'openSync');
```
Comment (robertsLando) [inline prelude/diagnostic.js:52]: What about iterating the prototype keys instead?

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/packer.js (hunk 17 above)
Comment (erossignon) [reply lib/packer.js:115]: Because of simplification in walker.js


 -    record[STORE_STAT] = await fs.stat(record.file);

Replace by; 
      const valueStat = await fs.stat(record.file);

      const value = {
        mode: valueStat.mode,
        size: valueStat.isFile() ? valueStat.size : 0,
        //
        isFileValue: valueStat.isFile(),
        isDirectoryValue: valueStat.isDirectory(),
        isSocketValue: valueStat.isSocket(),
        isSymbolicLinkValue: valueStat.isSymbolicLink(),
      };
      record[STORE_STAT] = value;
    }

File: prelude/diagnostic.js (hunk 18 above)
Comment (erossignon) [reply prelude/diagnostic.js:52]: No all are wrapped

**Review by robertsLando — APPROVED** [review]

File: .eslintrc [hunk 19]
```diff
@@ -20,10 +21,41 @@
     "import/prefer-default-export": "off",
     "camelcase": "off"
   },
-  "overrides": [{
-    "files": ["prelude/**/*"],
-    "rules": {
-      "strict": "off"
+  "overrides": [
```
Comment (leerob) [inline .eslintrc:24]: We turned a _lot_ of lint rules off here - why?

Comment (erossignon) [reply .eslintrc:24]: These are the minimum set of specific rules that make linter work for the test/*.js only as of today.
Fixing them, all would generate a massive commit and is outside the scope of this PR.
I suggest that we remove the rules one by one later one, in a different PR to put all test files in compliance with the default lint rules.

**Review by robertsLando — APPROVED** [review]

**Review by erossignon — COMMENTED** [review]
I have removed the commit related to extending lint to test file as well. I will reapply it in a separate PR.

File: lib/walker.js [hunk 20]
```diff
@@ -17,12 +18,20 @@ import {
...
-
+import {
+  detect,
+  visitor_MALFORMED,
+  visitor_NONLITERAL,
+  visitor_USESCWD,
+  visitor_SUCCESSFUL,
+} from './detector';
+
+const strictVerify = Boolean(process.env.PKG_STRICT_VER);
```
Comment (robertsLando) [inline lib/walker.js:34]: @erossignon Do you have added something on readme to document this?

File: prelude/bootstrap.js [hunk 21]
```diff
@@ -1051,8 +1079,8 @@ function payloadFileSync(pointer) {
 
   function statFromSnapshot(path_, cb) {
     var cb2 = cb || rethrow;
-    var path = normalizePath(path_);
-    // console.log("statFromSnapshot", path);
+    // don't follow
```
Comment (robertsLando) [inline prelude/bootstrap.js:1082]: the comment says don't follow but then you use a followPath, is this correct?

File: lib/walker.js [hunk 22]
```diff
@@ -636,16 +774,35 @@ class Walker {
...
+    }
+
     try {
-      record[STORE_STAT] = await fs.stat(record.file);
+      const valueStat = await fs.stat(record.file);
+
+      const value = {
+        mode: valueStat.mode,
+        size: valueStat.isFile() ? valueStat.size : 0,
+        //
```
Comment (robertsLando) [inline lib/walker.js:791]: Empty comment line

File: prelude/common.js [hunk 23]
```diff
@@ -153,7 +154,6 @@ exports.substituteDenominator = function substituteDenominator(f, denominator) {
 };
 
 exports.snapshotify = function snapshotify(file, slash) {
-  assert.strictEqual(file, normalizePath(file));
```
Comment (robertsLando) [inline prelude/common.js:156]: Why did you removed this?

File: lib/walker.js (hunk 20 above)
Comment (erossignon) [reply lib/walker.js:34]: @robertsLando  the strict strictVerify is to be used by developers of the pkg, not the end-user.
The primary goal is to make sure that we are storing content and state of real files, not symlinks.

The assertion done by strictVerify are quite expensive (following the link) and should not be enabled in production.

I take you point and add a comment at least about this line in the code, so that developers can understand what this is doing.

File: lib/walker.js (hunk 22 above)
Comment (erossignon) [reply lib/walker.js:791]: call it a separator between raw properties and calculated properties.

File: prelude/common.js (hunk 23 above)
Comment (erossignon) [reply prelude/common.js:156]: perfomance hit ?

Comment (erossignon) [reply prelude/common.js:156]: I'll add it back

File: lib/walker.js (hunk 20 above)
Comment (robertsLando) [reply lib/walker.js:34]: Ok right, but yeah we should document this somewhere as there could be many other devs in future in this project :) 

I'm thinking about a DEVELOPERS.md file where we add some docs to better explain how pkg works on the low level explaining what each file is used for and we could also add this process env var :)

Based on the PR you wrote here I think you have a very complete knowladge of how pkg works, would you like to submit a PR also for that document? I could help you there

File: lib/walker.js (hunk 22 above)
Comment (robertsLando) [reply lib/walker.js:791]: Oh ok so maybe add a comment to say that otherwise it seems something is missing

File: prelude/bootstrap.js [hunk 24]
```diff
@@ -1051,8 +1079,7 @@ function payloadFileSync(pointer) {
 
   function statFromSnapshot(path_, cb) {
     var cb2 = cb || rethrow;
-    var path = normalizePath(path_);
-    // console.log("statFromSnapshot", path);
+    var path = normalizePathAndFollowLink(path_);
```
Comment (robertsLando) [inline prelude/bootstrap.js:1082]: Just to be sure, is it ok to follow so? Was the comment you added here wrong?

File: prelude/common.js (hunk 23 above)
Comment (robertsLando) [reply prelude/common.js:156]: No problem for me if this doens't create possible bugs

Comment (erossignon) [reply prelude/common.js:156]: In fact the snapshotify method is only call here in 6 places

```` javascrript
export default function producer({ backpack, bakes, slash, target, symLinks }) {
  return new Promise((resolve, reject) => {
    if (!Buffer.alloc) {
      throw wasReported(
        'Your node.js does not have Buffer.alloc. Please upgrade!'
      );
    }

    const { prelude } = backpack;
    let { entrypoint, stripes } = backpack;
    entrypoint = snapshotify(entrypoint, slash);
    stripes = stripes.slice();

    const vfs = {};
    for (const stripe of stripes) {
      let { snap } = stripe;
      snap = snapshotify(snap, slash);
      if (!vfs[snap]) vfs[snap] = {};
    }

    const snapshotSymLinks = {};
    for (const [key, value] of Object.entries(symLinks)) {
      const k = snapshotify(key, slash);
      const v = snapshotify(value, slash);
      snapshotSymLinks[k] = v;
```` 

by design, symlink key and value have already been normalized, and snaps as well.
The only one that could potentiall not be normalized here is entrypoint... 

Due to the cost of this assert, I would move it elsewhere to verify that entrypoint is really normalized.

Comment (robertsLando) [reply prelude/common.js:156]: Ok agree :+1:

File: prelude/bootstrap.js (hunk 24 above)
Comment (erossignon) [reply prelude/bootstrap.js:1082]: Suppose that we have a path that is also a symlink.
When we asked for stat, shall we get the stat from the real file or for the link itself.
I ask myself this question. At one point I though we would need to separate the stat  from the link file from the real file. but it appears that we can avoid duplicating this information.
Looking at what the `fs.statSync` method does on a symlinked file we get the size of the realfile, the only difference is the nlink property that is set to 1.

image
> fs.statSync("a1.js")
Stats {
  dev: 2890017917,
  mode: 33206,
  nlink: 1,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: 4096,
  ino: 4503599628273627,
  size: 1729,
  blocks: 8,
  atimeMs: 1616854148464.44,
  mtimeMs: 1616503529688.595,
  ctimeMs: 1616503529688.595,
  birthtimeMs: 1616503255624.1414,
  atime: 2021-03-27T14:09:08.464Z,
  mtime: 2021-03-23T12:45:29.689Z,
  ctime: 2021-03-23T12:45:29.689Z,
  birthtime: 2021-03-23T12:40:55.624Z
}
> fs.statSync("a.js")
Stats {
  dev: 2890017917,
  mode: 33206,
  nlink: 1,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: 4096,
  ino: 4503599628273627,
  size: 1729,
  blocks: 8,
  atimeMs: 1616854148464.44,
  mtimeMs: 1616503529688.595,
  ctimeMs: 1616503529688.595,
  birthtimeMs: 1616503255624.1414,
  atime: 2021-03-27T14:09:08.464Z,
  mtime: 2021-03-23T12:45:29.689Z,
  ctime: 2021-03-23T12:45:29.689Z,
  birthtime: 2021-03-23T12:40:55.624Z
}

so using the stat stored in the real file snap safe some bytes as we emulate the behavior of fs.

So yes it is ok to resolve the link here,
(the comment was out dated)

Comment (robertsLando) [reply prelude/bootstrap.js:1082]: thanks for the explanation :) Resolved

File: lib/walker.js (hunk 22 above)
Comment (erossignon) [reply lib/walker.js]: I deleted the line, it was adding little purpose anyway

File: lib/walker.js (hunk 20 above)
Comment (erossignon) [reply lib/walker.js:41]: @robertsLando , please check my latest commit that contains the comment and tell me if it makes sense

Comment (erossignon) [reply lib/walker.js:41]: @robertsLando  => Good point a DEVELOPPERS.md file could be helpfull, ( in a separate PR)

File: lib/index.js [hunk 25]
```diff
@@ -484,7 +486,7 @@ export async function exec(argv2) {
   log.debug('Targets:', JSON.stringify(targets, null, 2));
 
   for (const target of targets) {
-    if (await exists(target.output)) {
+    if (fs.existsSync(target.output)) {
```
Comment (robertsLando) [inline lib/index.js:489]: fs not defined here ?

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 25 above)
Comment (erossignon) [reply lib/index.js]: Sorry I screw up , It was a bad idea to try  fix the "deprecated fs.exits" in this PR , I will address this in a separate PR.

Comment (hipstersmoothie) [reply lib/index.js]: I've fixed this in my TypeScript PR 😄

**Comment by robertsLando** [comment]
@leerob please give a look at this and merge when ready so we can go on with the typescript PR
